evm/src/cpu/kernel/asm/secp256k1/lift_x.asm

```diff
@@ -0,0 +1,817 @@
+// Returns y such that (x,y) is on Secp256k1 and y&1 = parity,
+// as well as a flag indicating whether such a y exists.
+%macro secp_lift_x
+    // stack: x, parity
+    %cubemodn_secp_base
+    // stack: x^3, parity
+    PUSH 7
+    // stack: 7, x^3, parity
+    %addmodn_secp_base
+    // stack: x^3+7, x, parity
+    DUP1
+    // stack: x^3+7, x^3+7, parity
+    %sqrt_secp_base
+    // stack: y, x^3+7, x, parity
+    SWAP1
+    // stack: x^3+7, y, parity
+    DUP2
+    // stack: y, x^3+7, y, parity
+    %squaremodn_secp_base
+    // stack: y^2, x^3+7, y, parity
+    EQ
+    // stack: sqrtOk, y, parity
+    SWAP2
+    // stack: parity, y, sqrtOk
+    DUP2
+    // stack: y, parity, y, sqrtOk
+    PUSH 1
+    // stack: 1, y, parity, y, sqrtOk
+    AND
+    // stack: 1 & y, parity, y, sqrtOk
+    EQ
+    // stack: correctParity, y, sqrtOk
+    DUP2
+    // stack: y, correctParity, y, sqrtOk
+    %secp_base
+    // stack: N, y, correctParity, y, sqrtOk
+    SUB
```

Comment:
Seems like this could give a non-canonical `-y` if `y=0`. But I guess it would be ignored since `sqrtOk` will be false in that case, correct?

Reply:
Oh I see, you also check `r != 0` in `ecrecover_input_check`